addons/website_slides/models/slide_channel.py

```diff
@@ -571,30 +669,94 @@ def action_add_member(self, **member_values):
-            to_join.message_subscribe(partner_ids=target_partners.ids, subtype_ids=[self.env.ref('website_slides.mt_channel_slide_published').id])
-            return slide_partners_sudo
-        return self.env['slide.channel.partner'].sudo()
+            # Contains indices of partners in 'existing'
+            existing_map_indices = dict((cid, list()) for cid in self.ids)
+            for pos, channel_partner in enumerate(existing):
+                existing_map[channel_partner.channel_id.id].append(channel_partner.partner_id.id)
+                existing_map_indices[channel_partner.channel_id.id].append(pos)
+
+            to_create_values = []
+            to_write_existing_indices = []
+
+            # Check if target_partners are already in channels to join.
+            # Create new members (as pending if invited, and enrolled otherwise)
+            for channel in to_join:
+                for partner in target_partners:
+                    try:
+                        ind = existing_map[channel.id].index(partner.id)
+                    except ValueError:
+                        channel_partner = dict(channel_id=channel.id, partner_id=partner.id, **member_values)
+                        to_create_values.append(channel_partner)
+                    else:
+                        to_write_existing_indices.append(existing_map_indices[channel.id][ind])
+
+            # Chatter subscription is done in create function override, if attendees are not pending.
+            created_partners_sudo = self.env['slide.channel.partner'].sudo().create(to_create_values)
+
+            # Enroll pending members and add them to channel(s) followers.
+            if to_write_existing_indices and not as_pending_member:
+                for existing_partner_index in to_write_existing_indices:
+                    existing[existing_partner_index].write({'is_pending': False})
+                enrollment_done = True
+                to_join.sudo().message_subscribe(
+                    partner_ids=[existing[ind].partner_id.id for ind in to_write_existing_indices],
+                    subtype_ids=[self.env.ref('website_slides.mt_channel_slide_published').id]
+                )
+
+        # (*) The logged partner self-enrolls in 'invite'-enrollment policy courses and subscribes to them.
+        if to_self_enroll and not as_pending_member:
+            self_enroll_channel_partners = self.env['slide.channel.partner'].sudo().search([
+                ('channel_id', 'in', to_self_enroll.ids),
+                ('partner_id', '=', logged_partner.id)
+            ])
+            self_enroll_channel_partners.write({'is_pending': False})
+            enrollment_done = True
+            to_self_enroll.sudo().message_subscribe(
+                partner_ids=[logged_partner.id],
+                subtype_ids=[self.env.ref('website_slides.mt_channel_slide_published').id]
+            )
+
+        return created_partners_sudo or enrollment_done
```

Comment:
This seems a bit overly-complicated.

Hopefully, if we move the "self-invited" logic back to "action_add_member", it should already clean it up?

I would like to keep something close to the previous logic, such as:

- build the list of "to_join" using "_filter_add_members"
- build a dict of existing channel_partner values
- initialize an empty recordset of "confirmed_invitation_patners_sudo"
- loop on channels
- loop on partners
- partner is not in existing -> add to create values like before (make sure to pass your "is_invitation_pending" field value based on our new "add_member_mode" parameter -> make it True when add_member_mode is 'invite')
- partner is existing AND partner invitation is pending AND "add_member_mode" is set to 'enroll' -> add this slide_channel_partner record to "confirm_invitation_partners_sudo" (<- the slide_channel_partner record should be present in the "existing" variable if you filter it properly)
- create values based on list of values to create, which will give you a recordset of "new_slide_channel_partner" records
- write "is_invitation_pending" = False on "confirmed_invitation_patners_sudo"
- if "add_member_mode" is 'enroll', subscribe both "new_slide_channel_partner" and "confirmed_invitation_patners_sudo" to the subtype
- return both "new_slide_channel_partner" and "confirmed_invitation_patners_sudo"

We can discuss this if it's unclear, but I think it should give you an idea.
Purpose is to have a clear flow.

Remember that your various use-cases should be thoroughly unit-tested as mentioned in a previous comment.

Reply:
Hello, the whole method code is reduced to 40 lines and feels much better. I also renamed a few variables to make it more readable (Not too much, I hope). I also saw is_member_pending_invitation and is_member_invitation_pending in previous comments so I kept is_member_invitation_pending. I must yet implement the tests.

About subscription, I must override the create function anyway to enroll to the channel the members created directly in the back end Tree view. Therefore, no need to subscribe both new_slide_channel_partners and confirmed_invitation_partners_sudo, only confirmed_invitation_partners_sudo.